### PR TITLE
Add readme documentation for demos

### PR DESCRIPTION
**Pull request by ninjasource**
Hi, I've recently worked through all the demos and took notes as I was going. I thought it may be useful to add those notes to this repo if appropriate. I'm not sure if this is the correct place to put them but adding a `README.md` file to the examples folder confuses cargo builds for some reason. I hope that this will be useful as I was not familiar with all the acronyms in the project and I think that newcomers will find this quick reference useful too.

### CODE REVIEW

**Comment by kalkyl** [comment]
Nice!

File: README.md [hunk 1]
```diff
@@ -51,3 +51,110 @@ at your option.
...
+
+The programmable peripheral interconnect (PPI) demo. The PPI allows peripherals to interact with each other without having to go through the CPU. Note that you need to choose a default feature in order for this demo to build. See above. This demo uses the Bluetooth RADIO peripheral as an example but does nothing special with Bluetooth itself so this is not the demo to learn about that capability.
+
+### pwm-demo (Digital Pins)
+
+The pulse width modulation demo. This demonstrates various PWM use cases by allowing the user to press buttons to change demo modes. This outputs PWM signals to the built in LEDs on the nRF52840-DK.
+
+### qdec-demo (Sensor Decoding)
+
+The quadrature decoder (QDEC) demo. This peripheral supports buffered decoding of quadrature-encoded sensor signals (typically used for mechanical and optical sensors). The demo reads a byte value from two input pins expected to contain qdec encoded data at a given frequency.
```
Comment (kalkyl) [inline README.md:136]:
```suggestion
The quadrature decoder (QDEC) demo. This peripheral supports buffered decoding of quadrature-encoded sensor signals (typically used for mechanical and optical sensors). The demo shows how to use a rotary encoder to trigger an interrupt and update a variable by the amount of rotation. (nRF52840 + Bourns PEC11R rotary encoder)
```

**Comment by ninjasource** [comment]
Thanks Henrik!

File: README.md [hunk 2]
```diff
@@ -51,3 +51,110 @@ at your option.
...
+
+I2C compatible Two-Wire Interface Master mode demo. This demo uses the TWIM peripheral to read and write 8 bytes of data to arbitrary addresses on a device that is connected to the I2C pins p0_30 and p0_31. It demonstrates error handling if the device does not respond properly (or it is not connected).
+
+### twis-demo (Digital Pins)
+
+I2C compatible Two-Wire Interface Slave mode demo. This demo uses the twis peripheral with rtic to listen for I2C signals which are exposed as events. The event handler reads data from the peripheral at the address specified.
+
+### wdt-demo (Timer)
+
+Watchdog timer demo. This demonstrates the how the entire device can be set to automatically reset if certain conditions are not met within a certain period of time. In this case you have to press all 4 buttons at least once within a 5 second period to prevent a reset. If you mash the buttons for a while you will encounter an 'attempt to subtract with overflow' panic at `main.rs:205` which is ultimately cleared by the watchdog timer. This is intended demo behaviour ;)
```
Comment (jonas-schievink) [inline README.md:160]: I think the individual demo descriptions should go into the demo's directory. That keeps all information related to a demo in a single directory and makes it easier to update.

Comment (ninjasource) [reply README.md:160]: Thanks, I’ll make that change when I’m back at my desk on Tuesday. Strange thing was that cargo got all confused when I put a dedicated README.md file in the examples folder. I’ll experiment.

Comment (ninjasource) [reply README.md]: Ok I've moved the demo related docs to its own file. Cargo was not confused, I was.

Comment (jonas-schievink) [reply README.md]: I meant that we should put a small readme into *each* demo, which would avoid the CI issues. That way, all information for a demo is located in its directory instead of being scattered.

Comment (ninjasource) [reply README.md]: Oh, I see. Some of the demos already have their own readme files. I wanted to add a quick reference guide targeted at developers unfamiliar with all the acronyms. Something to skim over when looking for a feature. For example. I was specifically looking for a bluetooth example (for which there are currently no dedicated examples) and I had to go through each demo to find it. I also didn't know how to run the demos and it's probably wasefull to repeat that information for each demo.

Comment (jonas-schievink) [reply README.md]: Yeah the "how to run" info is fine to just put in the main readme once.

Comment (ninjasource) [reply README.md]: Ok I will go ahead and create readme files for all the demos. Can I at least list the demos (and what their acronyms mean) in the main demo readme file? Things like ccm, comp, ecb, ppi, etc may be familiar to some but not to others.

Comment (jonas-schievink) [reply README.md]: Yeah that's fine. I just don't want to put all the details there since they might change.

Comment (ninjasource) [reply README.md]: Ok fair enough, cheers.

Comment (ninjasource) [reply README.md]: I've moved the bulk of the text to the relevant readme files for each demo and left a short title for each demo in their place in the form of a table in the readme file in the root of the examples folder. Should I move this detail back to the readme in the root of the project itself and delete the readme in the examples folder or leave it as-is?

Comment (ninjasource) [reply README.md]: Hi, I reread your comments and figured that you would have wanted me to add the detail to the main readme instead of creating a new one in the examples folder. I've gone ahead and done this. Sorry about the confusion.

Comment (ninjasource) [reply README.md]: > 
> 
> Yeah that's fine. I just don't want to put all the details there since they might change.

Hey I'm not sure if you have unsubscribed from this thread. Just a reminder to check out the latest if you have. I'd like to contribute more to this repo and picked up [#60](url) which is marked as a good first issue. I'd like to help out with code reviews if you have a backlog of PRs to work though and feel pressed for time. Or some other github issue, whatever helps.

David

File: README.md [hunk 3]
```diff
@@ -51,3 +51,69 @@ at your option.
...
+
+There are many ways to setup programming and debugging with an embedded device. Here we will describe how to do this on the nRF52840-DK using the [probe-rs](https://probe.rs/) set of tools. 
+
+### Once Off System Setup
+
+Install the cross compilation toolchain to target the device.
+```console
+$ rustup target add thumbv7em-none-eabihf
+```
+Install the tools to program and run on the device. See [probe-rs](https://github.com/probe-rs/probe-rs) for more details on other required dependencies.
```
Comment (jonas-schievink) [inline README.md:93]: probe-run is not part of probe-rs, and has no extra dependencies

File: README.md [hunk 4]
```diff
@@ -51,3 +51,69 @@ at your option.
...
+Each demo readme should contain instructions on how to run it. However, the information below describes the technologies used and can be used to troubleshoot your system setup. Run the demos from within their respective project directories. E.g. to run `ccm-demo`, you must be in the `nrf-hal/examples/ccm-demo/` directory and run `cargo run`.
+> Since the demos are stand-alone projects you would **NOT** typically run them with `cargo run --example xyz-demo` like some cargo projects are configured to do.
+
+There are many ways to setup programming and debugging with an embedded device. Here we will describe how to do this on the nRF52840-DK using the [probe-rs](https://probe.rs/) set of tools. 
+
+### Once Off System Setup
+
+Install the cross compilation toolchain to target the device.
+```console
+$ rustup target add thumbv7em-none-eabihf
```
Comment (jonas-schievink) [inline README.md:91]: This is not the right target for all chips

File: README.md [hunk 5]
```diff
@@ -51,3 +51,69 @@ at your option.
...
+```
+Install the tools to program and run on the device. See [probe-rs](https://github.com/probe-rs/probe-rs) for more details on other required dependencies.
+```console
+$ cargo install probe-run
```
Comment (jonas-schievink) [inline README.md:95]: We should not suggest probe-run as the default way to run the examples. The examples are not set up with probe-run, they still use cargo-embed (or OpenOCD + GDB, actually). I would be fine with moving them over to probe-run, but that should be done in a different PR.

File: examples/ccm-demo/README.md [hunk 6]
```diff
@@ -1,4 +1,8 @@
-# AES-CCM demo
+# The cipher block chaining - message authentication code (CCM) mode demo - Encryption
```
Comment (jonas-schievink) [inline examples/ccm-demo/README.md:1]: These titles seem too verbose

```suggestion
# CCM peripheral encryption demo
```

The acronym can be expanded in the text below

(same goes for the other readmes)

**Review by jonas-schievink — CHANGES_REQUESTED** [review]

File: README.md (hunk 5 above)
Comment (ninjasource) [reply README.md]: Good point. I have removed the references to probe-run and updated the main readme file accordingly.

File: README.md (hunk 4 above)
Comment (ninjasource) [reply README.md:90]: This a an example target. I have changed the wording above the command to give it more context.

File: README.md (hunk 3 above)
Comment (ninjasource) [reply README.md]: Removed

File: examples/spis-demo/README.md [hunk 7]
```diff
@@ -0,0 +1,11 @@
+# Serial peripheral interface slave (SPIS) - Audio
```
[the remark this answers is not in the document]
Comment (ninjasource) [reply examples/spis-demo/README.md]: That was a mistake, thanks.

File: examples/spi-demo/README.md [hunk 8]
```diff
@@ -1,11 +1,10 @@
-# spi-demo
-SPIM demonstation code.
-Connect a resistor between pin 22 and 23 on the demo board to feed MOSI directly back to MISO
-If all tests pass all four Led (Led1 to Led4) will light up, in case of error only at least one of the Led will remain turned off.
+# Serial peripheral interface master (SPIM) with EasyDMA demo - Audio
```
[the remark this answers is not in the document]
Comment (ninjasource) [reply examples/spi-demo/README.md]: Removed mistake.

**Review by thejpster — APPROVED** [review]

**Comment by thejpster** [comment]
@jonas-schievink are you content to clear your blocking bit on this now?

**Review by jonas-schievink — APPROVED** [review]
yeah, this seems fine

**Comment by ninjasource** [comment]
Hi Jonas and Jonathan, thanks for the approval. I "thought" that I addressed all the blocking things but I probably didn't click the right buttons for the PR to be handed over, my apologies! If there is anything still outstanding or things have gotten out of date since November 2020 I'd be happy to fix them in another PR. Just let me know.
David
